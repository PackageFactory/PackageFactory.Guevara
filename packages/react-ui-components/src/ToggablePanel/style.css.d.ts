export interface IStyleCss {
  'panel': string;
  'panel__headline': string;
  'panel__headline--noPadding': string;
  'panel__toggleBtn': string;
  'panel__contents': string;
  'panel--isOpen': string;
  'panel--condensed': string;
  'panel__contents--noPadding': string;
}

export const locals: IStyleCss;
