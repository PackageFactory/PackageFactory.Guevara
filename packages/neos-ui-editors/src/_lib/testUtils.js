import {PureComponent, Children} from 'react';
import PropTypes from 'prop-types';
import {SynchronousRegistry, SynchronousMetaRegistry} from '@neos-project/neos-ui-extensibility/src/registry';

const globalRegistry = new SynchronousMetaRegistry();

// i18n Registry
class FakeI18NRegistry extends SynchronousRegistry {
    translate(key) {
        return key;
    }
}
globalRegistry.set('i18n', new FakeI18NRegistry());

// Data Source DataLoader
export const MockDataSourceDataLoader = {
    _currentPromise: null,
    _currentPromiseResolveFn: null,

    // API in tests
    reset() {
        this._currentPromise = null;
    },

    resolveCurrentPromise(result) {
        this._currentPromiseResolveFn(result);
        return this._currentPromise;
    },

    // API of DataLoader
    resolveValue() {
        if (!this._currentPromise) {
            this._currentPromise = new Promise(resolve => {
                this._currentPromiseResolveFn = resolve;
            });
        }
        return this._currentPromise;
    }
};
globalRegistry.set('dataLoaders', new SynchronousRegistry());
globalRegistry.get('dataLoaders').set('DataSources', MockDataSourceDataLoader);

// wrapper
export class WrapWithMockGlobalRegistry extends PureComponent {
    static propTypes = {
        children: PropTypes.element.isRequired
    };

    static childContextTypes = {
        globalRegistry: PropTypes.object.isRequired,
        configuration: PropTypes.object.isRequired
    };

    getChildContext() {
        const configuration = {};
        return {globalRegistry, configuration};
    }

    render() {
        return Children.only(this.props.children);
    }
}
