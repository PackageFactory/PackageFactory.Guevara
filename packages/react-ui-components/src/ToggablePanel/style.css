.panel {
    composes: reset from './../reset.css';
    position: relative;
}

.panel__headline {
    composes: reset from './../reset.css';
    height: var(--goldenUnit);
    margin: 0;
    padding: 0 0 0 var(--spacing);
    line-height: var(--goldenUnit);
    font-weight: bold;
    font-size: var(--baseFontSize);
    border-top: 1px solid var(--brandColorsContrastDark);
    border-bottom: 1px solid var(--brandColorsContrastDark);
    display: flex;
}

.panel__headline--noPadding {
    padding: 0;
}

.panel__toggleBtn {
    composes: reset from './../reset.css';
}

.panel__toggleBtnTooltip {
  flex: 0 0 auto;
  margin-left: auto;
  z-index: 9;
}

.panel__contents {
    composes: reset from './../reset.css';
    padding: 0 var(--spacing);
    transition: .5s ease padding;

    .panel--isOpen & {
        padding-top: 5px;
        padding-bottom: 5px;
    }
    .panel--condensed.panel--isOpen & {
        padding-top: 0;
        padding-bottom: 0;
    }
}

.panel__contents--noPadding {
    padding: 0;
}
