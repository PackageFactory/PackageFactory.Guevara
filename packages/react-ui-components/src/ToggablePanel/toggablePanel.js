import React, {PureComponent, PropTypes} from 'react';
import Collapse from 'react-collapse';
import mergeClassNames from 'classnames';
import {Broadcast, Subscriber} from 'react-broadcast';

export default class ToggablePanel extends PureComponent {
    static propTypes = {
        /**
         * This prop controls if the contents are visible or not.
         */
        isOpen: PropTypes.bool,

        /**
         * Switches icon-open and icon-closed if set to true; can be used for
         * panels that close downwards such as the page structure tree.
         */
        closesToBottom: PropTypes.bool,

        /**
         * The children, ideally one Header and Contents component each.
         */
        children: PropTypes.any.isRequired,

        /**
         * The handler which will be called once the user toggles the contents.
         */
        onPanelToggle: PropTypes.func
    };

    static defaultProps = {
        isOpen: false
    };

    constructor(props, context) {
        super(props, context);

        this.toggle = this.toggle.bind(this);
        this.state = {
            isOpen: props.isOpen
        };
    }

    componentWillReceiveProps(newProps) {
        const {isOpen} = newProps;
        const isStateLess = Boolean(newProps.onPanelToggle);

        if (isOpen !== this.state.isOpen && !isStateLess) {
            this.setState({isOpen});
        }
    }

    render() {
        //
        // If the `onPanelToggle` prop is provided, the component will not
        // be using internal state, instead it will be controlled by the props.
        //
        const isStateLess = Boolean(this.props.onPanelToggle);
        const onPanelToggle = isStateLess ? this.props.onPanelToggle : this.toggle;
        const isOpen = isStateLess ? this.props.isOpen : this.state.isOpen;

        return (
            <StatelessToggablePanel
                {...this.props}
                isOpen={isOpen}
                onPanelToggle={onPanelToggle}
                >
                {this.props.children}
            </StatelessToggablePanel>
        );
    }

    toggle() {
        this.setState({isOpen: !this.state.isOpen});
    }
}

export class StatelessToggablePanel extends PureComponent {
    static propTypes = {
        /**
         * This prop controls if the contents are visible or not.
         */
        isOpen: PropTypes.bool,

        /**
         * Switches icon-open and icon-closed if set to true; can be used for
         * panels that close downwards such as the page structure tree.
         */
        closesToBottom: PropTypes.bool,

        /**
         * An optional className to render on the wrapper.
         */
        className: PropTypes.string,

        /**
         * The children, ideally one Header and Contents component each.
         */
        children: PropTypes.any.isRequired,

        /**
         * The handler which will be called once the user toggles the contents.
         */
        onPanelToggle: PropTypes.func.isRequired,

        /**
         * An optional css theme to be injected.
         */
        theme: PropTypes.shape({/* eslint-disable quote-props */
            'panel': PropTypes.string,
            'panel--isOpen': PropTypes.string
        }).isRequired/* eslint-enable quote-props */
    };

    static defaultProps = {
        isOpen: false
    };

    static childContextTypes = {
        onPanelToggle: PropTypes.func.isRequired
    };

    getChildContext() {
        return {
            onPanelToggle: this.props.onPanelToggle
        };
    }

    render() {
        const openedIcon = this.props.closesToBottom ? 'chevron-down' : 'chevron-up';
        const closedIcon = this.props.closesToBottom ? 'chevron-up' : 'chevron-down';

        const {children, className, theme} = this.props;
        const finalClassName = mergeClassNames({
            [className]: className && className.length,
            [theme.panel]: true,
            [theme['panel--isOpen']]: this.props.isOpen
        });

        return (
            <Broadcast channel="isPanelOpen" value={{isOpen: this.props.isOpen, openedIcon, closedIcon}}>
                <section className={finalClassName}>
                    {children}
                </section>
            </Broadcast>
        );
    }
}

export class Header extends PureComponent {
    static propTypes = {
        /**
         * The children which will be rendered within the header.
         */
        children: PropTypes.any.isRequired,

        /**
         * An optional css theme to be injected.
         */
        theme: PropTypes.shape({/* eslint-disable quote-props */
            'panel__headline': PropTypes.string,
            'panel__headline--noPadding': PropTypes.string,
            'panel__toggleBtn': PropTypes.string
        }).isRequired, /* eslint-enable quote-props */

        /**
         * Can be set to remove padding from the content area
         */
        noPadding: PropTypes.bool,

        /**
         * Static component dependencies which are injected from the outside (index.js)
         */
        HeadlineComponent: PropTypes.any.isRequired,
        IconButtonComponent: PropTypes.any.isRequired
    };

    static contextTypes = {
        onPanelToggle: PropTypes.func.isRequired
    };

    render() {
        const {
            HeadlineComponent,
            IconButtonComponent,
            children,
            theme,
            ...rest
        } = this.props;
        const {onPanelToggle} = this.context;

        const finalClassName = mergeClassNames([theme.panel__headline], {
            [theme['panel__headline--noPadding']]: this.props.noPadding
        });

        return (
            <Subscriber channel="isPanelOpen">{ ({isOpen, openedIcon, closedIcon}) =>
                <div aria-expanded={isOpen} {...rest}>
                    <HeadlineComponent
                        className={finalClassName}
                        type="h1"
                        style="h4"
                        >
                        {children}
                    </HeadlineComponent>
                    <IconButtonComponent
                        className={theme.panel__toggleBtn}
                        icon={isOpen ? openedIcon : closedIcon}
                        onClick={onPanelToggle}
                        />
                </div>
            }</Subscriber>
        );
    }
}

export class Contents extends PureComponent {
    static propTypes = {
        /**
         * An optional className to be rendered on the wrapping node.
         */
        className: PropTypes.string,

        /**
         * Can be set to remove padding from the content area
         */
        noPadding: PropTypes.bool,

        /**
         * The rendered children which can be toggled.
         */
        children: PropTypes.any.isRequired,

        /**
         * An optional css theme to be injected.
         */
        theme: PropTypes.shape({/* eslint-disable quote-props */
            'panel__contents': PropTypes.string,
            'panel__contents--noPadding': PropTypes.string
        }).isRequired/* eslint-enable quote-props */
    };

    static defaultProps = {
        theme: {}
    };

    render() {
        const {
            children,
            className,
            theme
        } = this.props;

        const finalClassName = mergeClassNames(theme.panel__contents, {
            [theme['panel__contents--noPadding']]: this.props.noPadding,
            [className]: className && className.length
        });

        return (
            <Subscriber channel="isPanelOpen">{ ({isOpen}) =>
                <Collapse isOpened={isOpen}>
                    <div className={finalClassName} aria-hidden={isOpen ? 'false' : 'true'}>
                        {children}
                    </div>
                </Collapse>
            }</Subscriber>
        );
    }
}
