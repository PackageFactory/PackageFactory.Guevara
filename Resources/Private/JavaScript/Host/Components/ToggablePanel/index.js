import React, {Component, PropTypes} from 'react';
import Transition from 'react-motion-ui-pack';
import mergeClassNames from 'classnames';
import Headline from 'Host/Components/Headline/';
import I18n from 'Host/Components/I18n/';
import Icon from 'Host/Components/Icon/';
import IconButton from 'Host/Components/IconButton/';
import style from './style.css';

export default class ToggablePanel extends Component {
    static propTypes = {
        title: PropTypes.string.isRequired,
        isOpened: PropTypes.bool,
        icon: PropTypes.string,
        className: PropTypes.string,
        headerClassName: PropTypes.string,
        children: PropTypes.node.isRequired
    }

    constructor(props) {
        super(props);

        this.state = {
            isOpened: props.isOpened
        };
    }

    componentWillReceiveProps(newProps) {
        const {isOpened} = newProps;

        if (isOpened !== this.state.isOpened) {
            this.setState({isOpened});
        }
    }

    render() {
        const {
            className
        } = this.props;
        const classNames = mergeClassNames({
            [className]: className && className.length,
            [style.panel]: true,
            [style['panel--isOpen']]: this.state.isOpened
        });

        return (
            <section className={classNames}>
                {this.renderHeader()}
                {this.renderContents()}
            </section>
        );
    }

    renderHeader() {
        const {
            title,
            icon,
            headerClassName
        } = this.props;
        const toggleIcon = this.state.isOpened ? 'chevron-up' : 'chevron-down';
        const className = mergeClassNames({
            [headerClassName]: headerClassName && headerClassName.length
        });

        return (
            <div className={className}>
                <Headline
                    className={style.panel__headline}
                    type="h1"
                    style="h4"
                    >
                    {icon ? <Icon icon={icon} padded="right" /> : null}
                    <I18n fallback={title} id={title} />
                </Headline>
                <IconButton
                    className={style.panel__toggleBtn}
                    icon={toggleIcon}
                    onClick={this.togglePanel.bind(this)}
                    />
            </div>
        );
    }

    renderContents() {
        return (
            <div>
                <Transition enter={{height: 'auto'}} leave={{height: 0}}>
                    {this.state.isOpened ? (
                        <div className={style.panel__contents}>
                            {this.props.children}
                        </div>
                    ) : null}
                </Transition>
            </div>
        );
    }

    togglePanel() {
        this.setState({
            isOpened: !this.state.isOpened
        });
    }
}
ToggablePanel.defaultProps = {
    isOpened: false
};
