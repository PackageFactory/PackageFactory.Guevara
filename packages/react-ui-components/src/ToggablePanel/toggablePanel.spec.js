import test from 'ava';
import {createShallowRenderer, createStubComponent} from './../_lib/testUtils.js';
import {render} from 'enzyme';
import React from 'react';
import ToggablePanel, {
    StatelessToggablePanel,
    Header,
    Contents
} from './toggablePanel.js';

const defaultThemeProps = {
    theme: {
        'panel': 'baseClassName',
        'panel--isOpen': 'isOpenClassName',
        'panel__headline': 'panelHeadlineClassName',
        'panel__contents': 'panelContentsClassName'
    }
};

const defaultProps = {
    ...defaultThemeProps,
    children: 'Foo children'
};
const shallow = createShallowRenderer(ToggablePanel, defaultProps);

const stateLessDefaultProps = {
    theme: {
        'panel': 'baseClassName',
        'panel--isOpen': 'isOpenClassName'
    },
    children: 'Foo children',
    onPanelToggle: () => null
};
const shallowStateLess = createShallowRenderer(StatelessToggablePanel, stateLessDefaultProps);

const HeadlineComponent = createStubComponent('HeadlineComponent');
const IconButtonComponent = createStubComponent('IconButtonComponent');

const defaultPropsFullRendering = {
    ...defaultThemeProps,
    HeadlineComponent,
    IconButtonComponent
};

test('<ToggablePanel/> should initialize with a state of {isOpen: props.isOpen}.', t => {
    const node = shallow();

    t.is(node.state('isOpen'), false);
});
test('<ToggablePanel/> should toggle the "isOpen" state when calling the instance "toggle" method.', t => {
    const node = shallow();

    t.is(node.state('isOpen'), false);

    node.instance().toggle();

    t.is(node.state('isOpen'), true);
});
test('<ToggablePanel/> should render a "StatelessToggablePanel" component.', t => {
    const node = shallow().find(StatelessToggablePanel);

    t.is(node.length, 1);
});
test('<ToggablePanel/> should propagate the rest of the passed props to the "StatelessToggablePanel" component.', t => {
    const node = shallow({className: 'testClassName'}).find(StatelessToggablePanel);

    t.is(node.prop('className'), 'testClassName');
});
test('<ToggablePanel/> should propagate the "isOpen" prop to the the "StatelessToggablePanel" component in case a "onPanelToggle" prop was provided.', t => {
    const node = shallow({isOpen: true, onPanelToggle: () => null}).find(StatelessToggablePanel);

    t.is(node.prop('isOpen'), true);
});
test('<ToggablePanel/> should propagate the "onPanelToggle" prop to the the "StatelessToggablePanel" component in case it was provided.', t => {
    const onPanelToggle = () => null;
    const node = shallow({isOpen: true, onPanelToggle}).find(StatelessToggablePanel);

    t.is(node.prop('onPanelToggle'), onPanelToggle);
});
test('<ToggablePanel/> should propagate the "isOpen" state to the the "StatelessToggablePanel" component in case no "onPanelToggle" prop was provided.', t => {
    const wrapper = shallow();
    const node = wrapper.find(StatelessToggablePanel);

    t.is(node.prop('isOpen'), wrapper.state('isOpen'));
});
test('<ToggablePanel/> should propagate the "toggle" instance method to the the "StatelessToggablePanel" component in case no "onPanelToggle" prop was provided.', t => {
    const wrapper = shallow();
    const node = wrapper.find(StatelessToggablePanel);

    t.is(node.prop('onPanelToggle'), wrapper.instance().toggle);
});

test('<StatelessToggablePanel/> should render only the "panel" className of the theme in any case.', t => {
    const wrapper = shallowStateLess(null);

    t.truthy(wrapper.childAt(0).hasClass(stateLessDefaultProps.theme.panel));
    t.falsy(wrapper.childAt(0).hasClass(stateLessDefaultProps.theme['panel--isOpen']));
});
test('<StatelessToggablePanel/> should render both the "panel" and "panel--isOpen" className of the theme in case the "isOpen" prop is truthy.', t => {
    const wrapper = shallowStateLess({isOpen: true});

    t.truthy(wrapper.childAt(0).hasClass(stateLessDefaultProps.theme.panel));
    t.truthy(wrapper.childAt(0).hasClass(stateLessDefaultProps.theme['panel--isOpen']));
});
test('<StatelessToggablePanel/> should render the "className" prop if provided.', t => {
    const wrapper = shallowStateLess({isOpen: true, className: 'FooClassName'});

    t.truthy(wrapper.childAt(0).hasClass('FooClassName'));
});
test('<StatelessToggablePanel/> should render its propagated children.', t => {
    const wrapper = shallowStateLess();

    t.truthy(wrapper.html().includes('Foo children'));
});

const fullyRenderToggablePanel = isOpen => {
    return render(<ToggablePanel isOpen={isOpen} {...defaultPropsFullRendering}>
        <Header {...defaultPropsFullRendering}>My Header</Header>
        <Contents {...defaultPropsFullRendering}>My my Content</Contents>
    </ToggablePanel>);
};

const getHeaderDomNode = panelDomNode =>
    panelDomNode.children('section').children('div').first();

const getContentsDomNode = panelDomNode =>
    panelDomNode.children('section').children('div').first().next();

test('open ToggablePanel should render a <ToggablePanel.Header/> with an "aria-expanded" attribute of "true".', t => {
    const wrapper = fullyRenderToggablePanel(true);

    t.truthy(getHeaderDomNode(wrapper).attr('aria-expanded'));
});

test('closed ToggablePanel should render a <ToggablePanel.Header/> with an "aria-expanded" attribute of "false".', t => {
    const wrapper = fullyRenderToggablePanel(false);

    t.is(getHeaderDomNode(wrapper).attr('aria-expanded'), 'false');
});

test('<ToggablePanel.Header/> should render a "HeadlineComponent" and "IconButtonComponent" component.', t => {
    const wrapper = fullyRenderToggablePanel(false);

    t.is(wrapper.find('[data-component-name=HeadlineComponent]').length, 1);
    t.is(wrapper.find('[data-component-name=IconButtonComponent]').length, 1);
});
test('<ToggablePanel.Header/> should render a "HeadlineComponent" with the themes "panel__headline" className.', t => {
    const wrapper = fullyRenderToggablePanel(false);

    t.is(getHeaderDomNode(wrapper).find('[data-component-name=HeadlineComponent]').attr('class'), defaultThemeProps.theme.panel__headline);
});
test('<ToggablePanel.Header/> should render all children within the "HeadlineComponent".', t => {
    const wrapper = fullyRenderToggablePanel(false);
    t.is(getHeaderDomNode(wrapper).text(), 'My Header');
});

test('<ToggablePanel.Header/> should render a "IconButtonComponent" with an icon of "chevron-down" if the "isOpen" context is falsy.', t => {
    const wrapper = fullyRenderToggablePanel(false);
    t.is(getHeaderDomNode(wrapper).find('[data-component-name=IconButtonComponent]').attr('icon'), 'chevron-down');
});

test('<ToggablePanel.Header/> should render a "IconButtonComponent" with an icon of "chevron-up" if the "isOpen" context is truthy.', t => {
    const wrapper = fullyRenderToggablePanel(true);
    t.is(getHeaderDomNode(wrapper).find('[data-component-name=IconButtonComponent]').attr('icon'), 'chevron-up');
});

test('<ToggablePanel.Contents/> should render the themes "panel__contents" className.', t => {
    const wrapper = fullyRenderToggablePanel(true);
    t.truthy(getContentsDomNode(wrapper).find('[aria-hidden]').hasClass(defaultThemeProps.theme.panel__contents));
});

test('<ToggablePanel.Contents/> should render a falsy "aria-hidden" attribute if the "isOpen" context is truthy.', t => {
    const wrapper = fullyRenderToggablePanel(true);
    t.is(getContentsDomNode(wrapper).find('[aria-hidden]').attr('aria-hidden'), 'false');
});
test('<ToggablePanel.Contents/> should render a truthy "aria-hidden" attribute if the "isOpen" context is falsy.', t => {
    const wrapper = fullyRenderToggablePanel(false);
    t.is(getContentsDomNode(wrapper).find('[aria-hidden]').attr('aria-hidden'), 'true');
});
