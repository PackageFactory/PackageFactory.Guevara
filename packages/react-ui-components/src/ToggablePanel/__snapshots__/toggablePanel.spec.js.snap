// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Contents/> should render correctly. 1`] = `
<Collapse
  fixedHeight={-1}
  isOpened={false}
  keepCollapsedContent={false}
  onHeightReady={[Function]}
  style={Object {}}
>
  <div
    aria-hidden="true"
    className="panelContentsClassName"
  >
    Foo children
  </div>
</Collapse>
`;

exports[`<Header/> should render correctly. 1`] = `
<div
  aria-expanded={false}
>
  <Component
    className="panelHeadlineClassName"
    type="h1"
  >
    Foo children
    <Component
      className="panelToggleBtnClassName"
      icon="closed-icon-prop"
      onClick={[Function]}
      tooltipLabel="Toggle the panel"
      tooltipPosition="left"
    />
  </Component>
</div>
`;

exports[`<StatelessToggablePanel/> should render correctly. 1`] = `
<section
  className="baseClassName"
>
  Foo children
</section>
`;

exports[`<ToggablePanel/> should render correctly. 1`] = `
<StatelessToggablePanel
  HeadlineComponent={[Function]}
  IconButtonComponent={[Function]}
  isOpen={true}
  onPanelToggle={[Function]}
  theme={
    Object {
      "panel": "baseClassName",
      "panel--isOpen": "isOpenClassName",
      "panel__contents": "panelContentsClassName",
      "panel__headline": "panelHeadlineClassName",
    }
  }
>
  <Header
    HeadlineComponent={[Function]}
    IconButtonComponent={[Function]}
    closedIcon="chevron-down"
    isPanelOpen={true}
    key="header"
    openedIcon="chevron-up"
    theme={
      Object {
        "panel": "baseClassName",
        "panel--isOpen": "isOpenClassName",
        "panel__contents": "panelContentsClassName",
        "panel__headline": "panelHeadlineClassName",
      }
    }
  >
    My Header
  </Header>
  <Contents
    HeadlineComponent={[Function]}
    IconButtonComponent={[Function]}
    isPanelOpen={true}
    key="contents"
    theme={
      Object {
        "panel": "baseClassName",
        "panel--isOpen": "isOpenClassName",
        "panel__contents": "panelContentsClassName",
        "panel__headline": "panelHeadlineClassName",
      }
    }
  >
    My my Content
  </Contents>
</StatelessToggablePanel>
`;
